### PR TITLE
Support to auto correct a lint

### PR DESCRIPTION
**Pull request by brauliobo**
FIXME: broke monkey patch at lib/scss_lint/sass/tree.rb

### CODE REVIEW

**Comment by brauliobo** [comment]
the intention was not the current API, so the change was quite small to achieve the apply behaviour

**Comment by brauliobo** [comment]
change the monkey path to be less intrusive and all tests pass now

**Comment by sds** [comment]
Hey @brauliobo,

There's a lot going on in this pull request without much documentation. From what I see, you are trying to do four separate things:

* Replace `scss-lint`'s monkey patching of `Node#children` with a separate `linter_children` call
* Add support for automatically fixing lints? (the `--apply` flag name isn't very descriptive here. Other tools usually use the name `--auto-correct`)
* Split `MergeableSelector` into separate linters (Why? An explanation of the purpose for this would be useful, as it introduces a backwards incompatibility with developers who configure `MergeableSelector`)
* Add a new `GroupableSelector` linter (again, with no documentation--please add documentation to the [linter README](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md))

Please split these out into separate pull requests with clear descriptions of the problems they are trying to solve. This allows us to merge them individually without holding up the whole set due to problems/concerns with one of them.

Furthermore, you've probably noticed that the lint checks are not passing. Please ensure all RuboCop-related warnings are addressed before submitting a pull request. Thanks!

**Comment by brauliobo** [comment]
@sds, nice, gonna follow your advices, hope I can find some time soon.

**Comment by brauliobo** [comment]
@sds applied your suggestions

The first two suggestions is the this PR, as the Sass::Tree fix is core dependency for auto correcting
Just created the #535 and #536 for the last two.

Still there is Rubocop issues, sorry for this, I'm quite not used to it, will get used to it and fix the issues.

**Comment by brauliobo** [comment]
fixed rubocop warnings

File: lib/scss_lint/cli.rb [hunk 1]
```diff
@@ -54,9 +54,12 @@ def act_on_options(options)
     end
 
     def scan_for_lints(options, config)
+      files = FileFinder.new(config).find(options[:files])
+
       runner = Runner.new(config)
-      runner.run(FileFinder.new(config).find(options[:files]))
+      engines = runner.run(files)
```
Comment (sds) [inline lib/scss_lint/cli.rb]: This is not an ideal approach. Calling `run` should not return a list of engines. The runner should auto-correct if auto-correction has been called for. There are a lot of other issues with the `Runner` class that are unrelated to your change, but it's purpose is to be the central coordinator for the tool. The `CLI` class is simply intended to parse CLI options and hand those off to the `Runner`.

File: lib/scss_lint/cli.rb [hunk 2]
```diff
@@ -169,6 +172,15 @@ def report_lints(options, lints, files)
       end
     end
 
+    def auto_correct(engines)
+      runner.lints.each do |lint|
+        lint.auto_correct.call if lint.auto_correct
+      end
+      engines.each do |engine|
+        puts engine.tree.to_scss
```
Comment (sds) [inline lib/scss_lint/cli.rb]: I don't think this works the way you think it does. Calling `to_scss` is likely to change more of the original source code than you originally intended due to the quirks of the Sass engine. 

Furthermore, using `puts` to output the results prevents us from controlling the file we output to via p[`--out` flags](https://github.com/brigade/scss-lint/pull/273). A better solution would simply update the file at the end of the run.

However, the big show-stopper here is using `to_scss`, which will butcher the original source tree.

**Comment by sds** [comment]
Hey @brauliobo, thanks for the pull request. Really appreciate the effort to add support for something that I know a lot of users are looking for.

Sadly, I'm concerned with the general approach here. Relying on `to_scss` is not a viable option, as it will re-serialize the entire source code tree. As I've already explained in https://github.com/brigade/scss-lint/issues/343#issuecomment-70279869, other projects (such as `rubocop`) benefit from a better parse tree which can actually be changed at runtime and re-serialized without touching the rest of the source code. We don't have that flexibility with the `sass` gem here, so any solution will need to implement that from scratch (a huge undertaking, IMO).

**Comment by sds** [comment]
Closing due to lack of response. Feel free to reopen if you can come up with a new approach!

**Pull request closed without merging**